Update PID Tuning

joshuabardwell opened the pull request:
This update is based on expected changes to be released in the 1.9.1 code version. Examples include the new soft features, adding TPA to PIDC#1 and 2, and removing the I term from the TPA calculation. I believe that these changes have all been accepted into the master and are slated for release in the 1.9.1 code version.

JohnOCFII commented:
Great documentation!  Thanks for taking the time to create it.

digitalentity commented on docs/PID tuning.md:
```diff
@@ -4,30 +4,70 @@ Every aspect of flight dynamics is controlled by the selected "PID controller".
+with you, some trial and error is required to find the best performing PID settings. PID settings are also somewhat dependent on a pilot's preferences, so even if someone else has your exact same craft, you may prefer slightly different PID settings than them.
 
-A video on how to recognise and correct different flight problems caused by PID settings is available here:
+A video on how to recognise and correct different flight problems caused by PID settings is available [here](https://www.youtube.com/watch?v=YNzqTGEl2xQ). A detailed thread on tuning a mini-quad for acro-style FPV flight is [here](http://www.rcgroups.com/forums/showthread.php?t=2439428).
```
You have created an excelent guide there on rcgroups, but I think this guide should probably be part of cleanflight docs.

digitalentity replied:
Oops, my bad. This IS mostly the same guide.

joshuabardwell replied:
More or less, since the same guy wrote it. I tried to keep the level of detail lower for the CF docs, and people who want to do a deeper dive can go to the RCGroups thread.

fakatwork replied:
Just a little precision while we are at it, the "rewrite" PIDC is also cycletime aware. Maybe we could add a line in the same vein as lux or harakiri to precise that changing the "looptime" setting shouldn't impact the PID tuning too much whith this one. 

Alternatively, I'd like if we could make it clear that for all other controllers excepted those three, changing the "looptime" setting has a clear impact and will likely imply to retune the PID parameters.

borisbstyle commented:
Nice documentation!

smitty97 commented:
Excellent writeup! I wish I had this 4 months ago.

A few things that have helped me understand things better:  I like how you say to reduce P by a percentage, not just up/down "a few clicks".  It wasn't until I found [this page and interactive demo](http://www.gperco.com/2014/05/quadcopter-stability-and-neural-networks.html#canvasjump) that I realized the terms are used in multiplication, and that the difference between entering 1.1 and 1.3 translates to a bigger change than the difference between 5.1 and 5.3.  I think this should be stressed a bit more for noobs like me.  (Actually, if the CF configurator also had a 'slider' like the demo, that would help too)

Another resource I found is the [OpTune Calculator](http://op.thatwebplace.com.au/optune.php).  (Instructions [here](https://forums.openpilot.org/topic/39941-tuning-guide-for-noobs-and-new-builds-optune/))  It seems like a good method, you set I & D to zero, and find the P that has steady oscillation at hover.  Plug them into this calculator and it spits out new PIDs based on (I think) the [ZN method](https://en.wikipedia.org/wiki/Ziegler%E2%80%93Nichols_method) as well as some other factors for aggressive vs stable.  I'm not sure how the numbers translate for the controllers in CF (should we divide I by 10?) but it'd be nice to have something like this for the CF controllers.

henn1001 commented:
OPTune is indeed very interesting and it should be indeed the standard ZN method. Just a manual(and probably safer) aproach to autotuning. But what i don´t get is how he calculates the Tcrit / Oszilation frequency? As far as i know these should be different for every Quadcopter depending on the motorisation and the especially the size?

smitty97 commented:
@henn1001 Not sure, he's probably making some assumptions based on the fact that you can only choose certain frame sizes and flying styles.

ctzsnooze commented:
ZN method doesn't care about the device, type of quad, or whatever is being controlled, only the relationship between PID and wobble. Problem is not knowing Ku, the amplitude of the undamped oscillation, in any useful way, with quads.  However we can reduce P until undamped ongoing oscillation is barely present, and use P value that as an estimate of Ku then reduce by 0.6 to get a working P value.  Then our next problem is the units of D, for which the scaling is different in each controller and the units are arbitrary. 

In practice these methods are not particularly useful. 

Better is to set D and I to zero, then look for the value of P at which point there are 2-3 wobbles only (often a good idea to hand test, or string quad up between legs of chair, rather than real flight test), then add D until those wobbles are visually adequately damped.  Blackballing helps fine tune it.

Another important consideration is resistance to external force with zero I.  Too much P and in acro while hand-held you find it very strongly resisting any push, and trying to oscillate, too little P and it feels miushy.  After a bit of experience, 'enough' is not so difficult to know.  :-)

jaapp commented on docs/PID tuning.md:
```diff
@@ -4,30 +4,70 @@ Every aspect of flight dynamics is controlled by the selected "PID controller".
+
+If P is set too high, the craft will rapidly oscillate/shake as it continually overshoots its target. Excess P gain is the single easiest thing to troubleshoot in multirotor tuning because the oscillations that result are so distinct. Excess P gain is demonstrated in [this video](https://www.youtube.com/watch?v=RMQXSB5-6bE). 
+
+A good guideline for setting P gain is to increase P until oscillations start to occur during normal control input (e.g. moderate turns and forward flight). Then decrease P to between 50% and 80% of that value. Another way of determining the oscillation threshold is to riase the throttle quickly from a hover and look/listen for oscillations during the punchout. Often times, espeically if flying l ...
```
"riase" shoud be "raise"

Very well written up documentation!

ctzsnooze replied:
Higher P gains are not the primary way of setting responsiveness to control inputs.  RCrate and PitchRoll rate have a much wider range of usable values and a bigger influence on stick responsiveness.  P gains are  vital for keeping the quad stable, i.e. the ability to resist external forces like gusts of wind.  

Maybe change that line to:   "Higher P gain helps the craft better resist external forces, such as gusts of wind, from pushing it off its intended angle in flight, and tends to give sharper responses to control inputs."  Then next sentence.. 

"One of the most common *mistakes* made by novice tuners is having too much P"  

Maybe then

"Often this causes visible wobbles and oscillations after sudden stick inputs."  

then "Even if these oscillations.... "

ctzsnooze commented on docs/PID tuning.md:
```diff
@@ -4,30 +4,70 @@ Every aspect of flight dynamics is controlled by the selected "PID controller".
+
+Higher P gain tends to give a craft a sharper response to control inputs, but one of the most common problems made by novice tuners is having excess P. Even if oscillations aren't visible, there can be micro-oscillations that make their way to the motors, that can result in decreased flight time and hotter motors. In general, it's probably a good idea for a novice tuner to err slightly on the sid ...
+
+**The I term**. Ideally, the P term would be all that is needed to have a perfect controller. In reality, the P term is always a little bit "behind the curve", because the copter takes some time to respond, and by the time it responds, the situation has changed. The I term tracks the cumulative error accrued by the P term and attempts to compensate for it. 
```
The need for the I term is not because the P term is a bit late.  The P term being a bit delayed, in association with the momentum of the quad, is why we get overshoot wobble.  Dealing with overshoot wobble is why we have the D term.  The I term is not there because the P term is a bit late.

The I term is actually there because the P term must always be set slightly below the force required to fully correct an error.  If the P term resulted in a force that was was greater than the force required to completely correct an error, it would always 'overdo' the correction, causing uncontrollable oscillation.  

As a result, the P term alone will not fully correct for persistent offsets, especially small persistent offsets.  This is where the I term comes in.  It accumulates up these small errors over time and compensates for them.  

A good I analogy might be when a quadcopter in level mode but has a battery that has slipped a bit so the weigh distribution is now off-centre, such that it tends to persistently lean to one side.  A P response from the imbalance in weight will attempt to correct the lean, but cannot be quite strong enough - on its own - to entirely solve the problem.  The I term accumulates that error over time, and gradually builds enough force to entirely correct the lean and get the craft much closer to nearly perfectly level.  Likewise I is essential to deal with  slow variable external forces - like wind resistance at greater speed - which would otherwise to change the intended angle of the craft as we add throttle.

jaapp commented:
Because I found the documentation very sparse (until this PR) I've been studying the code the last week to get a better sense of the PID control loop. And I've come away with a slightly different understanding of the D-term than I read everywhere else on the forums. Including in this PR.

Let's look at how the delta value is calculated in pidRewrite:
delta = RateError - lastError[axis];

And lets take a craft which is perfectly level, but gets a control input to roll to the right with 10dps. Right at the beginning (the first code loop where the rc signal is interpreted) the RateError will be pretty high since the craft still needs to start going into the roll. Which results in a massively high D-term spike(with the same mathematical sign as the P value), also known as the D-term kick.

As soon as the RateError seems to be accelerating the fastest towards a delta of 0, the D-Term will spike again but this time with a reversed mathematical sign and thus dampening the P value. Which is there to prevent overshoot.

Everywhere I read about the D-term as a dampening factor, but the way I see it, it is an accelerator as well. To call the D-term just a dampener might not give the best understanding.

To me this accelerating effect seems to be a vital flying characteristic. Having a high D-gain could main that the craft is way more responsive, and not only preventing oscillation. It gives you an extra reason why you might want to increase D in some scenario's.

borisbstyle commented:
@jaapp This PR is about the documentation ;)

jaapp commented:
@borisbstyle yes I know. Maybe I didn't make my point clear enough. My input was that we might have to stop calling the D-term a dampening factor, as it is confusing. It did confuse me for a while. From documentation and forum-talk I could never fully understand what it was doing, as the blackbox trace told me a different story.

So my suggestion is to stop calling it a brake, or a dampener. Because this is not the full story and (in my case at least) leads to confusion,

borisbstyle commented:
It's not dampening, but derivative or differential. Not sure who came up with dampening?

ledvinap commented:
@japp: D-term is mathematically equivalent to damping factor in spring-mass system (https://en.wikipedia.org/wiki/Harmonic_oscillator). P-term is equivalent to spring stiffness (because of this it's almost impossible to tune PID without D - tuning is possible only because quite small aerodynamic damping, increasing P only increases frequency).

ledvinap commented:
BTW: You can quite accurately model single axis as cylinder on bearings that is connected by P-term spring to setpoint and D-term viscous damper either to ground or setpoint ...

ctzsnooze commented:
Exactly.  D is always damping (reducing) the rate of change.  Like a car shock absorber.  

If a car spring has been compressed slowly, and then suddenly released, D sets the maximum rate of change of length of the spring, slowing the rate and amplitude of the oscillation.  In this situation D opposes P and can be well described as 'damping' the response of the spring.

However, if an external force is applied to the spring, D and P both oppose the force together, just as a hard shock absorber makes the car spring seem even harder over bumps.   With just the right combination of D and P we get both a firmer suspension and minimal oscillation.

When we change the setpoint (length of the spring), D goes against the spring attaining the new position.  That's why if in the original LuxFloat if you were to increase D past the optimal point, the quad became obviously sluggish in response to stick inputs.  Below the optimal point, P itself was the limiting factor in the rate of change to the new position (magnitude of P vs rotational inertia being dominant)

If we were to quickly and momentarily push against a stable hovering quad along one axis, the initial response from both P and D should be in-phase.  We should see this in blackbox. Both working to keep it still

But if we apply a stick input, D and P will be out of phase, because the new error signal will increase Pterm, which will start to change the angle of the quad, and D will oppose that change.  Only the Dterm kick is in-phase.

So yes, describing D as a damping factor is a fairly good descriptor.  And the kick factor reduces the damping factor during stick inputs, but not otherwise.

jaapp commented:
@borisbstyle At least we agree on that it is not a dampening factor. And indeed, this documentation does not use the term(but it links to a forum post that does). It speaks exclusively about the breaking and 'backing off' the P-term and doesn't explain that it is also an accelerator. Or am I misunderstanding the code?

There is major confusion of the role of the d-term. As is demonstrated right here where you and @ledvinap seem to contradict each other in two consecutive posts.

What I'm sharing with you is a major misunderstanding I had when I started to PID tune using blackbox, because nobody seemed to address the fact the the dterm is also an accelerator. I had to learn about the kick factor the hard way.

ctzsnooze commented:
No it is not an accelerator, Dterm always inhibits acceleration.  Dterm kick is an accelerator though.

jaapp commented:
@ctzsnooze But dterm kick is a part of the dterm.

borisbstyle commented:
"We are trying to achive damping effect with Derivative" 
D is not damping, but D has damping effect :D. It just happens both words are starting with D :D
Damping itself would be expressed by gain like for example 3 or 6dB, which is actually done with filters

ctzsnooze commented:
True but Dterm kick is not a significant accelerator in what we see so far, it only is enough to attenuate the normal Dterm responses.  If if could be amplified, then it would be an accelerator, a pulse occurring with every step change.  But for now it does not accelerating.

It's also true that Dterm is not a classical viscous energy absorber.  If there is a delay between Dterm output and response, and there always is, and if the Dterm gain is set very high, which sometimes happens, then there can be Dterm induced oscillations.  Classical viscous energy absorbers can never increase oscillation, only can they reduce it.

So it's true that Dterm is not a classical damping system.

However most of the effects it has on quadcopter behaviour is very similar to a classical damper, and the mathematical modelling is based on the principle of derivative damping.

The derivative is the mathematics of calculating the slope of the control value, but damping is the process of using that calculated derivative value to dampen the natural oscillation tendency of a control system.

So our D factor is at the same time a derivative (thats how its calculated) and a damping factor (in application and practice).

ctzsnooze commented:
PS I usually hand-launch quadcopter with moderate throttle.  Immediately before launch I often push briefly, quickly and transiently against each axis.  I get confirmation that the PID system is OK this way.  

It's amazing to think that, in acro mode, Dterm gives immediate resistance, proportional to how fast we push against the angle; Pterm gives medium term resistance proportional to how far I move it away from the set angle, and Iterm gives longer term steadily growing resistance to persistent fixed angle offsets after both the other terms are either zero or falling.

Very clever!  :-)

jaapp commented:
@ctzsnooze thank you, I finally understand why nobody seems to recognize my point. I think I've been overestimating the effect of the dterm kick. The amplitude of the kick might be very high, but it's probably too short-lived to have any accelerating effect.

borisbstyle commented:
Here some real examples. Roll axis during sharp stick move with and without derivative kick. The P jump is still quite high actually!


Without Derivative kick (current luxfloat)
![luxfloat_old](https://cloud.githubusercontent.com/assets/10757508/8651453/34e8e8b6-2979-11e5-910c-c5359e7f66df.jpg)


With Derivative kick
![Uploading luxfloat_new.jpg…]()

borisbstyle commented:
It definatelly seems to be accelerating the move :dancer:

ctzsnooze commented:
Yeah the derivative kick during the onset phase of the stick input has almost made the  Dterm effects almost zero.  As a result PIDsum is more rapidly responding to the input.

And in the offset phase everything is faster.

It might be nice to see the stick input, total PID and the gyro trace separately?

jaapp commented:
@borisbstyle interesting! There is no scaling on the graph, so hard to judge the effect. But but the PIDsum seems to have much larger amplitudes. I guess the jury is still out then.

Anyway, maybe this discussion is better to be had somewhere else? My goal was to give some feedback on the documentation, but it seems that I've stepped on something which might deserve it's own thread.

borisbstyle commented:
@jaapp The scaling is exactly the same. Same quad with same stable cycletime and logging and same zoom on the blackbox

I will make a final verdict and comparisation with provided data. Just give me some more time.

ctzsnooze commented:
Am amazed at how quickly and deeply our understanding of this really fundamental stuff has progressed!

borisbstyle commented:
I will also show that smoothed rccommand without derivative kick produces
different effect on pids than the old luxfloat approach with gyrorate.
Op 13 jul. 2015 16:36 schreef "ctzsnooze" <notifications@github.com>:

> Am amazed at how quickly and deeply our understanding of this really
> fundamental stuff has progressed!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cleanflight/cleanflight/pull/1089#issuecomment-120950711>
> .
>

joshuabardwell commented:
I don't have a deep understanding of PID systems, but I think (thought?) I understood them conceptually. And I remember from Calc 1 what a derivative is. When f(x) is constant, dy/dx=0. When f(x) is increasing or decreasing, dy/dx is non-zero. The D term is always subtracted from the P term so that the D term inhibits the P term when D is nonzero.

This is why I thought of the D term as a dampener or brake. Given these parameters, I don't see how D can ever accelerate the P term, at least in an ideal situation. But we have seen that D term kick is actually the same sign as the P term, not the opposite sign. I guess I need to do a little studying to understand how this can happen.

joshuabardwell commented:
On a side note--what is the correct way to incorporate these edits on Github? Should I edit the file associated with this commit, or should I do a new commit and pull request? Please describe in very, very basic terms, as I'm still learning how to use Github. Thanks.

ledvinap commented:
@joshuabardwell : dy/dt can be both positive (y rising) and negative (y falling). The sign is chosen so that D-term tried to bring its input rate of change to zero.
D-term and P-term are not in phase (D-term has 90deg phase lead), all combination of signs are possible and normal.

joshuabardwell commented:
>D-term and P-term are not in phase (D-term has 90deg phase lead), all combination of signs are possible and normal

I guess that's the key. Because D term is out of phase, it can be rising when P term is falling, if P term was rising previously.

ctzsnooze commented:
BorisB:  "I will also show that smoothed rccommand without derivative kick produces
different effect on pids than the old luxfloat approach with gyrorate."

Maybe... from my reading, the kick effect will still be there, but the spike won't... that's the theoretical prediction.

borisbstyle commented:
check my logs in #1110 with finaly very smoothed rccommand with no added
delay compared to before

BorisB: "I will also show that smoothed rccommand without derivative kick
produces
different effect on pids than the old luxfloat approach with gyrorate."

Maybe... from my reading, the kick effect will still be there, but the
spike won't... that's the theoretical prediction.

—
Reply to this email directly or view it on GitHub
<https://github.com/cleanflight/cleanflight/pull/1089#issuecomment-121073812>
.

joshuabardwell commented:
I have incorporated these edits, as well as changing all references to 1.9.1 to 1.10. I updated the file in my pid-documentation branch. Do I need to submit another pull request?

joshuabardwell commented:
@hydra Sorry to bother, but is this ready to go? Or have I messed up the PR in some way? With 1.10 RC now released, I would like to make sure I am not the holdup.

martinbudden commented:
@joshuabardwell , I'm in the process of going through all the documentation PRs and getting them merged in. It would be great to get this one merged in. Can you update it and remove reference to the now defunct PID controllers 4 & 5 and also rebase and resolve the merge conflicts? Once that is done I'll make best efforts to get this merged in quickly.
